quickjs.c

```diff
@@ -50959,20 +50879,293 @@ static const JSClassShortDef js_weakref_class_def[] = {
     { JS_ATOM_WeakRef, js_weakref_finalizer, NULL }, /* JS_CLASS_WEAK_REF */
 };
 
+typedef struct JSFinRecEntry {
+    struct list_head link;
+    JSValue obj;
+    JSValue target;
+    JSValue held_val;
+    JSValue token;
+} JSFinRecEntry;
+
+typedef struct JSFinRecData {
+    struct list_head entries;
+    JSContext *ctx;
+    JSValue cb;
+} JSFinRecData;
+
+static void delete_finrec_weakref(JSRuntime *rt, JSFinRecEntry *fre)
+{
+    JSWeakRefRecord **pwr, *wr;
+
+    pwr = get_first_weak_ref(fre->target);
+    for(;;) {
+        wr = *pwr;
+        assert(wr != NULL);
+        if (wr->kind == JS_WEAK_REF_KIND_FIN_REG && wr->u.fin_rec_entry == fre)
+            break;
+        pwr = &wr->next_weak_ref;
+    }
+    *pwr = wr->next_weak_ref;
+    js_free_rt(rt, wr);
+}
+
+static void js_finrec_finalizer(JSRuntime *rt, JSValue val)
+{
+    JSFinRecData *frd = JS_GetOpaque(val, JS_CLASS_FINALIZATION_REGISTRY);
+    if (frd) {
+        struct list_head *el, *el1;
+        list_for_each_safe(el, el1, &frd->entries) {
+            JSFinRecEntry *fre = list_entry(el, JSFinRecEntry, link);
+            delete_finrec_weakref(rt, fre);
+        }
```

Comment:
> Is this an accidentally-quadratic (well, O(n*m)) thing? Perhaps it's better to use a doubly linked list / circular queue so removal is O(1) and this loop stays O(n)?
> 
> I suppose the counterargument is that it makes every JSObject one almost-always-unused pointer bigger. And you know, that's actually a pretty good counterargument.
> 

So here is the problem: you have a finalization registry where you did this:

```
finReg.register(target);
finReg.register(target, undefined, target);
```

The 2nd line is dumb, but allowed. The script reaches the end and the runtime is being destroyed. The finalization registry will go first.

Now as we destroy it, `target` will have a refcount of 1, but 2 weak ref entries.

Because we added the one with the token last it will be the first one to go. As its refcount reaches 0 it will be freed and we'll be left hanging on to freed memory as we continue iterating that list.

Since weakrefs don't seem to be very used I think keeping it simple, as in as small overhead as possible is a good compromise. A similar pattern is used in reset_weak_ref I reckon for a similar reason.

Thanks for the review!

> As an aside: you could use plain list_for_each here since the list isn't mutated. Not that I object to using the _safe variant, just mentioning it.

Yeah, I kinda left it that way because it doesn't hurt...